[Bootcamp T124004534] Better Transformer fastpath diagnostics

mikekgfb opened the pull request:
Summary:
update the boolean logic for TransformerEncoder => https://www.internalfb.com/code/fbsource/[0da9d46c97ca76432303e55492b1e8b4131c781a]/fbcode/caffe2/torch/nn/modules/transformer.py?lines=206-212%2C228-230
and TransformerEncoderLayer => https://www.internalfb.com/code/fbsource/[0da9d46c97ca76432303e55492b1e8b4131c781a]/fbcode/caffe2/torch/nn/modules/transformer.py?lines=410-417%2C432-436

according to the decision logic => https://www.internalfb.com/code/fbsource/[0da9d46c97ca76432303e55492b1e8b4131c781a]/fbcode/caffe2/torch/nn/modules/activation.py?lines=1059-1091%2C1103-1109

Differential Revision: D37666451

jbschlosser commented on torch/nn/modules/transformer.py:
```diff
@@ -228,13 +249,20 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
+            if torch.overrides.has_torch_function(tensor_args):
+                why_not_fast_path = "some Tensor argument has_torch_function"
+            elif not (output.is_cuda or 'cpu' in str(output.device)):
+                why_not_fast_path = "src is neither CUDA nor CPU"
```
nit: should this mention src or output?

jbschlosser reviewed (approved):
LGTM!

ftyan123 committed:
Better Transformer fastpath diagnostics (#81013)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/81013

update the boolean logic for TransformerEncoder => https://www.internalfb.com/code/fbsource/[0da9d46c97ca76432303e55492b1e8b4131c781a]/fbcode/caffe2/torch/nn/modules/transformer.py?lines=206-212%2C228-230
and TransformerEncoderLayer => https://www.internalfb.com/code/fbsource/[0da9d46c97ca76432303e55492b1e8b4131c781a]/fbcode/caffe2/torch/nn/modules/transformer.py?lines=410-417%2C432-436

according to the decision logic => https://www.internalfb.com/code/fbsource/[0da9d46c97ca76432303e55492b1e8b4131c781a]/fbcode/caffe2/torch/nn/modules/activation.py?lines=1059-1091%2C1103-1109

Test Plan:
Pass unit tests in sandcastle
Export to github
Pass all Github CI/CD

Reviewed By: zrphercule

Differential Revision: D37666451

fbshipit-source-id: 639eb7023ca1f929ae3cbcb3f6590242dbf0f7e6
